packages/core/src/contrib/Constraints.ts

```diff
@@ -971,41 +936,42 @@ export const overlap1D = (
 };
 
 /**
- * Return the bounding box (as 4 points) of an axis-aligned box-like shape given by `center`, width `w`, height `h` as an object with `minX, maxX, minY, maxY`.
+ * Input: A rect-like shape.
+ * Output: A bbox corresponding to the rect.
  */
-export const bbox = (center: VecAD, w: VarAD, h: VarAD): any => {
-  const halfWidth = div(w, constOf(2.0));
-  const halfHeight = div(h, constOf(2.0));
-  const nhalfWidth = neg(halfWidth);
-  const nhalfHeight = neg(halfHeight);
-  // CCW: TR, TL, BL, BR
-  const pts = [
-    [halfWidth, halfHeight],
-    [nhalfWidth, halfHeight],
-    [nhalfWidth, nhalfHeight],
-    [halfWidth, nhalfHeight],
-  ].map((p) => ops.vadd(center, p));
+export const bbox = ([t, s]: [string, any]): BBox => {
+  if (!(isRectlike(t) || isLinelike(t))) {
+    throw new Error(
+      `Bbox expected a rect-like or line-like shape, but got ${t}`
+    );
+  }
 
-  const rect = {
-    minX: pts[1][0],
-    maxX: pts[0][0],
-    minY: pts[2][1],
-    maxY: pts[0][1],
-  };
+  // TODO: Compute the bbox of the line in a nicer way
 
-  return rect;
-};
+  // initialize center, w, and h depending on whether the input shape is line-like or rect/square-like
+  const center = isLinelike(t)
+    ? ops.vdiv(ops.vadd(s.start.contents, s.end.contents), constOf(2))
+    : s.center.contents;
+
+  const w =
+    t === "Square"
+      ? s.side.contents
+      : isLinelike(t)
+      ? max(absVal(sub(s.start.contents[0], s.end.contents[0])), constOf(2))
+      : s.w.contents;
+
+  const h =
+    t === "Square"
+      ? s.side.contents
+      : isLinelike(t)
+      ? max(absVal(sub(s.start.contents[1], s.end.contents[1])), constOf(2))
+      : s.h.contents;
 
-/**
- * Return the bounding box (as 4 segments) of an axis-aligned box-like shape given by `center`, width `w`, height `h` as an object with `top, bot, left, right`.
- */
-export const bboxSegs = (center: VecAD, w: VarAD, h: VarAD): any => {
```

Comment:
Not totally sure about this design decision. I try to avoid catch-all utility functions as they do unnecessary computation, which gets added to the hot path of optimization. This amount of extra computing is probably fine, but in the future, I would prefer to split up functions into several small ones that can be called by constraint/objectives as needed.

Reply:
I refactored `BBox` into an immutable class that lazily computes `corners` as needed.

Reply:
I replaced the ternaries with if-else.